src/charm.py

```diff
@@ -205,6 +207,110 @@ def _check_kf_profiles(self, interfaces):
 
         return kf_profiles
 
+    def main(self, event) -> None:
+        """Main entry point for the Charm."""
+        try:
+            self._check_container_connection()
+            self._check_model_name()
+            self._check_leader()
+            interfaces = self._get_interfaces()
+            kf_profiles = self._check_kf_profiles(interfaces)
+            self.handle_ingress(interfaces)
+        except CheckFailed as e:
+            self.model.unit.status = e.status
+            return
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            self.logger.error(traceback.format_exc())
+            self.unit.status = BlockedStatus("kubernetes resource creation failed")
+            return
+        self.handle_ingress(interfaces)
+        kf_profiles = list(kf_profiles.get_data().values())[0]
+        self.profiles_service = kf_profiles["service-name"]
+        try:
+            self._update_layer()
+        except CheckFailed as e:
+            self.model.unit.status = e.status
+            return
+        self.model.unit.status = ActiveStatus()
+
+    def _get_sidebar_config(self) -> Dict:
+        try:
+            current_configmap = self.k8s_resource_handler.lightkube_client.get(
+                ConfigMap, name=self._context["configmap_name"]
+            )
+            old_sidebar_config = json.loads(current_configmap.data["links"])
+        except ApiError as e:
+            self.logger.warning(
+                f"PROBLEM during configmap retrieval {e}. USING BASE CONFIG"
+            )
+            old_sidebar_config = json.loads(BASE_SIDEBAR)
+        return old_sidebar_config
+
+    def _update_sidebar_config(self, links_data: Dict) -> None:
+        self._context["links"] = json.dumps(links_data)
+        try:
+            self.k8s_resource_handler.context = self._context
+            self.k8s_resource_handler.template_files = [
+                self._resource_files["config_maps"]
+            ]
+            self.k8s_resource_handler.apply()
+        except ApiError as e:
+            self.unit.status = BlockedStatus(
+                f"kubernetes resource creation failed with {e}"
+            )
```

Comment:
another comment without a good place... 

Do we handle anywhere here the order of the sidebar links?  I didn't see it but may have missed that.  I don't know if relations are guaranteed in any order, and it would be odd to have a kubeflow sidebar that is different randomly every deploy.  Maybe we should include in the charm an option "core links always on top" that, if enabled (and enable it by default), will always keep any of the core links that have been provided through relations (notebooks, kfp, automl, ...) at the top of the bar and in the same order as kubeflow 1.6 had.  

For any of those core links, if they are not related then they don't show in the sidebar.  The feature just orders what **is** related

Reply:
Sorry, I later noticed the sorting in the relation_changed events.  I still think this would be a good addition (I don't think alphabetical by text is the best link prioritisation), but I was wrong above as they do have a deterministic order

Reply:
Instead of alphabetical order I added new field to config data named as position and I sort based on that. Your idea is hard to implment imho. What are the base links ? Is it kfp or what is it? The sort will be quite strange. I think :)

Reply:
By default the kfp has the lowest position.

Reply:
@misohu and I discussed and we will bring up the best way to order this with the team in a future meeting.  For now, any order is still good and shouldn't block the rest of this code from being reviewed.  We can modify as needed later